designdocs/user-handles.md

```diff
@@ -0,0 +1,323 @@
+* Query the chain to check which values are available, check if the handle with the suffix is available.
+* Choose an available value and attempt to claim it by submitting the full handle with the suffix to the chain.
+* Chain validates the handle and suffix and maps the handle to the ```msa_id```.
+* Chain maintains a reverse mapping from handle to msa for offchain use.
+
+### Chain Steps
+
+* Every 100 blocks (e.g., when block_number % 100 == 1), update the current seed by taking the Merkle root of the previous block. Keep the current and one previous seed so that two seeds, s0 and s1, are part of the chain state at any given time.
+* Provide an RPC to get the current seed.
+* Provide an RPC to get the suffix options for a given handle.
+* When a user submits a handle, check the numeric suffix for availability and that it honors range defined for suffix window.
+* Store ```msa_id``` to ```handle``` mapping and ```handle``` to ```msa_id``` mapping.
+
+### Handling Race Conditions
+
+There is an exceedingly small chance that two users may attempt to claim the same handle and suffix combination within the same block. In such cases, the transaction will fail to create the MSA, and the app will receive a failure for the transaction rather than the expected MsaHandleCreated event.
+
+## Sequence Diagram
+
+```mermaid
+sequenceDiagram
+    participant Wallet
+    actor User
+    participant App
+    participant RPC as RPC Node
+    participant Chain as Frequency
+    User->>App: Enter desired handle
+    App->>RPC: get_suffix_options(handle)
+    RPC->>Chain: Query current seed
+    Chain-->>RPC: Return seed
+    RPC->>RPC: Compute suffix options
+    RPC->>Chain: check_handle_availability(options)
+    Chain-->>RPC: Return subset of options
+    RPC-->>App: Return subset of options
+    App->>User: Present suffix options
+    User->>App: Choose desired suffix
+    App->>Wallet: Send signup data
+    Wallet->>Wallet: Perform 2FA if first attempt
+    Wallet-->>App: Return signed payload
+    App->>RPC: create_msa_handle(..., handle, suffix)*
+    RPC->>Chain: Submit transaction
+    Chain->>Chain: Check validity and availability
+    Chain-->>App: MsaHandleCreated event with handle
+    App->>User: Proceed with setup
+    User->>App: Request to get MSA ID for handle
+    App->>RPC: get_msa_id_for_handle(handle)
+    RPC->>Chain: Query handle mapping state (State Query MSA<->Handle)
+    Chain-->>RPC: Return handle mapping state (State Query MSA<->Handle)
+    RPC-->>App: Return MSA ID for handle
+    User->>App: Request to get handle for MSA ID
+    App->>RPC: get_handle_for_msa_id(msa_id)
+    RPC->>Chain: Query MSA ID mapping state (Reverse Index Handle<->MSA)
+    Chain-->>RPC: Return MSA ID mapping state (Reverse Index Handle<->MSA)
+    RPC-->>App: Return handle for MSA ID
+```
+
+## Storage
+
+* **HandleSuffixMin**: This storage value (```u32``) will keep track of the minimum suffix value allowed for a given handle.
```

Comment:
Do suffixes (used to distinguish accounts with same base handle) really need to be as large as 4294967295 (u32)? or can they be 65535 (u16)?

Reply:
I see.  There are ~294,282 Zhang Wei and ~44,935 John Smiths.

Reply:
We can decide of types but in general the window of min max suffix range will move around in that space